FIX (peripheral): add support for instantiating multiple peripherals of same type

T-K-233 opened the pull request:
changelog:changed
<!--
First, please ensure that the title of your PR is sufficient to include in the next changelog.
Refer to https://github.com/ucb-bar/chipyard/releases for examples and feel free to ask reviewers for help.

Then, make sure to label your PR with one of the changelog:<section> labels to indicate which section
of the changelog should contain this PR's title:
  changelog:added
  changelog:changed
  changelog:fixed
  changelog:removed

If you feel that this PR should not be included in the changelog, you must still label it with
changelog:omit

Provide a brief description of the PR immediately below this comment, if the title is insufficient -->


Change how peripherals are included, so that user can instantiate multiple peripheral devices of same type in chip config.

**Related PRs / Issues**:
<!-- List any related PRs/issues here, if applicable -->

<!-- choose one -->
**Type of change**:
- [x] Bug fix
- [ ] New feature
- [ ] Other enhancement

<!-- choose one -->
**Impact**:
- [x] RTL change
- [ ] Software change (RISC-V software)
- [ ] Build system change
- [ ] Other

<!-- must be filled out completely to be considered for merging -->
**Contributor Checklist**:
- [x] Did you set `main` as the base branch?
- [x] Is this PR's title suitable for inclusion in the changelog and have you added a `changelog:<topic>` label?
- [x] Did you state the type-of-change/impact?
- [x] Did you delete any extraneous prints/debugging code?
- [x] Did you mark the PR with a `changelog:` label?
- [ ] (If applicable) Did you update the conda `.conda-lock.yml` file if you updated the conda requirements file?
- [ ] (If applicable) Did you add documentation for the feature?
- [ ] (If applicable) Did you add a test demonstrating the PR?
<!-- Do this if this PR is a bugfix that should be applied to the latest release -->
- [ ] (If applicable) Did you mark the PR as `Please Backport`?

abejgonzalez commented on generators/chipyard/src/main/scala/config/fragments/PeripheralFragments.scala:
```diff
@@ -26,15 +28,20 @@ class WithBootROM extends Config((site, here, up) => {
   case PeripheryUARTKey => Seq(
-    UARTParams(address = 0x54000000L, nTxEntries = 256, nRxEntries = 256, initBaudRate = baudrate))
+    UARTParams(address = address, nTxEntries = 256, nRxEntries = 256, initBaudRate = baudrate))
+})
```
I think an API that makes more sense to me is: `class WithUARTOverride(no: Int, ...)` where the `no` is the index in the `PeripheryUARTKey` `Seq` to override.

abejgonzalez replied:
Or if you want to truely override the entire key I would have two APIs:

- one called `WithUARTOverride` that takes a Seq of tuples (enforced to be size of 2 - where the 1st element in the tuple is the address and the 2nd is the baudrate) that gets unrolled into multiple UARTParams.
- another that is the additive uart (same as the current `WithUART`)

jerryz123 replied:
I actually prefer the style here, as it matches how the WithNCores behaves

T-K-233 replied:
Ideally, we should get rid of the default UART (i.e. when user does not specify `WithUART`, we shouldn't include a UART in the SoC). 

I could not find where it adds the default UART though, so this is a hack to override the default parameter.

abejgonzalez replied:
The default UART is given here: 

https://github.com/ucb-bar/chipyard/blob/44abc91cc147d77651d1d2c1b5585fcc35727ba3/generators/chipyard/src/main/scala/config/AbstractConfig.scala#L48

We have the CY config have a UART by default so that we can boot Linux easily in SW-RTL sims.

I actually think we can change the default implementation of `WithUART` here:

https://github.com/ucb-bar/chipyard/blob/44abc91cc147d77651d1d2c1b5585fcc35727ba3/generators/chipyard/src/main/scala/config/fragments/PeripheralFragments.scala#L35-L38

To essentially be the "override" config (i.e. generalize it to override all the `Seq[UARTParams]` given a `Seq` of tuples) and renaming it as such `WithUARTOverride`. Then we can keep the current implementation of `WithUART` seen here (where it's additive and like `WithN...Cores`) and just move it to the `PeripheralFragments.scala` file.

jerryz123 replied:
We want to keep the default UART. There's hardly a good reason to build a SoC without a UART....

The default UART config should be close to "universal".. that is, rarely should users need to change the address/params of that UART.

Additional WithUART flags can add more UARTs at other addresses to the system..


We should have an *additional* WithUARTOverride fragment for special cases when the default UART settings are not correct ... I don't think that fragment will be used very frequently, but it should be available at least.

T-K-233 commented:
Revive this PR thread:
- Removed `WithUARTOverride` per requested.
- Added several other helpful peripheral config fragments, which are used in the tapeout class chip.
- Minor addition on ignoring the .vscode/ folder.

T-K-233 commented:
Reopen the PR
